### PR TITLE
PM sender RPC handling

### LINKED ISSUE #633: Use PM Sender in RPC
**Is your feature request related to a problem? Please describe.**
This issue assumes we've implemented `Sender`'s functionality, but haven't yet connected it to the video transcoding flow.

For this issue we want to incorporate `Sender` into the video flow.

**Describe the solution you'd like**
- Create a `Sender` per broadcasting session with the appropriate params.
- Use it to construct PM tickets.
- Attach a ticket to each outgoing segment.

_Note:_ this issue does not handle any failure conditions B might have relating to PM, such as low deposit, etc.

**Describe alternatives you've considered**
Nothing to add.

**Additional context**
Nothing to add.

### PR DESCRIPTION
**Pull request by yondonfu**
**What does this pull request do? Explain your changes. (required)**

Implements PM sender RPC message handling for a broadcaster.

**Specific updates (required)**

- Add a PM sender to the LivepeerNode
- Use TicketParams included in OrchestratorInfo messages to start local PM sessions
- Create Ticket payments which are attached to the header of segment messages sent to orchestrators

**How did you test each of these updates (required)**

I wrote additional unit tests.

**Does this pull request close any open issues?**

Fixes #633 

**Checklist:**

- [ ] README and other documentation updated
- [x] Node runs in OSX and devenv
- [x] All tests in `./test.sh` pass

### CODE REVIEW

**Comment by yondonfu** [comment]
Ready for review

File: server/mediaserver.go
Comment (eladmallel) [inline server/mediaserver.go:203]: what's the meaning behind the name `protoParams`? why not stick with `ticketParams`?

Comment (eladmallel) [reply server/mediaserver.go:203]: Oh, I see that `proto` is referring to protobuf. What's the value in defining this variable versus just using `tinfo.TicketParams` ?

File: server/mediaserver.go
Comment (eladmallel) [inline server/mediaserver.go]: Perhaps it's simpler to just add `Recipient` into the `pm.TicketParams` struct?

File: server/rpc.go
Comment (eladmallel) [inline server/rpc.go]: why do we log this error specifically whereas with other errors we return with no logging?

File: server/mediaserver.go
Comment (j0sh) [inline server/mediaserver.go:229]: Curious why it's done this way -- is there a reason for returning an ID to the PM session itself, rather than the PM session itself (or an interface thereof with a `CreateTicket` function)? That way the PM session doesn't have to be stored externally, and we only have to add one field to the connection context. Is there a reason to retain the session forever on the sender side?

File: server/mediaserver.go
Comment (yondonfu) [reply server/mediaserver.go:203]: I defined the additional variable to avoid having multiple instances of `tinfo.TicketParams.<FieldName>` when converting into a `pm.Ticket` struct. However, I'm thinking of just using a helper here:

```
params := net.FromProtoTicketParams(tinfo.TicketParams)
```

WDYT?

File: server/rpc.go
Comment (yondonfu) [reply server/rpc.go]: I forgot to remove this log - I don't think it is necessary because we wrap the error returned by `genPayment()` and log it [here](https://github.com/livepeer/go-livepeer/pull/662/files#diff-44bb03d888e11d6429223a7b0e55220dR477)

File: server/mediaserver.go
Comment (yondonfu) [reply server/mediaserver.go:229]: The intent was to to keep the `pm` package unaware of how it would eventually be used and for as much of the PM related logic to be handled out of the box by the `pm` package. We initially considered storing a `pm.Sender` per `BroadcastSession`, however, we realized we would still eventually need some way to synchronize all these `pm.Sender` instances so that they all have an up to date view of the remaining account deposit so that no `pm.Sender` instance overspends because another instance sent out many winning tickets. So, rather than having the `pm` package user handling deposit view synchronization we can use a single `pm.Sender` instance across each `BroadcastSession`. In this case, we still need a way the single `pm.Sender` instance to keep track of the current `senderNonce` for a set of ticket params, so we use an internal session within the `pm.Sender` that is referenced by its sessionID (which is just `recipientRandHash`).

I realize that the design probably could be simplified since the user of the `pm` package already has a notion of session, but we were interested in keeping the `pm` package as reusable as possible without any assumption that the user already has a notion of a session. The current design of the `pm` package would allow users to use it to send one-to-many out of the box without the user having to implement its own notion of a session. Definitely interested to hear if you have any thoughts on how this design could be improved though!

Comment (j0sh) [reply server/mediaserver.go:229]: Ah I see. The synchronization could still happen under the hood within `pm` as tickets are created. How will the account synchronization happen? I'm assuming via `TicketBroker` -- which probably needs to be threadsafe anyway.

I'm not sure what's meant by the "user having its own notion of a session". The user needs to store a session ID (or many); it could also store a PM session (or many) that's defined as part of the `pm` package. Storing sessions on the user eliminates that duplicate layer of bookkeeping within the `Sender` and makes it less explicitly stateful. Either way, there is some tracking required from the user.

Also, by storing sessions proper, we don't need to retain session information indefinitely (do we?) as sessions would be GC'd once the connection terminates. Right now, we accumulate sessions in memory without cleaning them up, which is not ideal for a long-running service process. Eventually this will exhaust the available memory.

The one reason I can see the current setup being useful is if the `Sender` needs a way to scan over all the (active) sessions. But I don't really see a clear need for this at the moment.

Perhaps it'd help me to understand how the fully-formed `Sender` interface is intended to work? The comments still indicate there are some TODOs.

To accommodate the current setup, I was thinking something like:

```
interface PMSession {
  CreateTicket(...) // syncs deposit view under the hood via broker?
}
func NewPMSession(broker TicketBroker, params TicketParams) PMSession
...
struct BroadcastSession {
  PMSession session
  ...
}
```

Comment (yondonfu) [reply server/mediaserver.go:229]: > How will the account synchronization happen?

One possibility is for `Sender` to maintain a running view of the account's deposit (which might be different from the on-chain deposit) if the recipient notifies the sender of winning tickets prior to redemption. This would also be compatible with your suggestion of creating a `PMSession` object per each set of ticket params - the `PMSession` object could hold a reference to the `Sender` object and would query its running view of the account's deposit when creating a ticket.

> I'm not sure what's meant by the "user having its own notion of a session". 

I meant that in our case there is already a `BroadcastSession` object used for streams so one approach would be to eliminate the notion of a PM session and simply add additional fields to the `BroadcastSession` object to track PM related information. However, it would be nice for users of the package that do not have the equivalent of a `BroadcastSession` object to not have to worry about defining their own session objects unless they absolutely need to.

> Also, by storing sessions proper, we don't need to retain session information indefinitely (do we?) as sessions would be GC'd once the connection terminates. Right now, we accumulate sessions in memory without cleaning them up, which is not ideal for a long-running service process. Eventually this will exhaust the available memory.

Great point - I really like this benefit of your suggestion.

> Perhaps it'd help me to understand how the fully-formed Sender interface is intended to work?

The original intent was for the global `Sender` object to handle all the state related to sending tickets including tracking internal sessions and `BroadcastSession` would just use `sessionID` to tell the `Sender` object which session it cares about without it knowing anything about the details of the session. Later on, we could add support for receiving new ticket params mid stream and we could replace the existing sessionID in `BroadcastSession` with the new sessionID when this happens.

Overall, I like your suggestion - what do you think about tackling it as a part of the refactor period post-MVS in the interest of not increasing the scope of this PR beyond its core intent (which is just integrating the PM sender into the node and handling RPC messages)?

File: server/mediaserver.go
Comment (j0sh) [reply server/mediaserver.go:203]: The helper is fine, but I wouldn't put it in the `net` package since that's become the de facto place for shared type/interface declarations to avoid import cycles.

My feeling is it's best to define methods closest to where they're used, so putting the helper in `mediaserver.go` (or  somewhere in `server`) is probably better.

**Comment by j0sh** [comment]
> However, it would be nice for users of the package that do not have the equivalent of a BroadcastSession object to not have to worry about defining their own session objects unless they absolutely need to.

I don't think there is a requirement for a user-supplied "broadcast session" . Users still have to track PM sessions regardless. The question here is whether they track session IDs, or a pm-provided `PMSession`.

> what do you think about tackling it as a part of the refactor period post-MVS in the interest of not increasing the scope of this PR

Sure, but if anything, it'd actually reduce the scope of the PR, since there won't be a need for a `sync.Map` or Sender-side session tracking, and one fewer field in the BroadcastSession (just the `PMSession`  rather than both the `Sender` + `SessionID`).

File: server/mediaserver.go
Comment (angyangie) [reply server/mediaserver.go:229]: When would a user not have the a `BroadcastSession` (or the equivalent, though not sure what that refers to)?

**Review by eladmallel — APPROVED** [review]
LGTM pending merging master into this branch and resolving conflicts!

**Comment by j0sh** [comment]
Just a PSA, a `git rebase master` rather than a `git merge master` will avoid those merge messages.

**Comment by yondonfu** [comment]
> Just a PSA, a git rebase master rather than a git merge master will avoid those merge messages.

Woops...yeah forgot to rebase instead of merge.

> Sure, but if anything, it'd actually reduce the scope of the PR, since there won't be a need for a sync.Map or Sender-side session tracking, and one fewer field in the BroadcastSession (just the PMSession rather than both the Sender + SessionID).

Good points. Though I think there will still be a bit of refactoring and code moved around so my preference is still to do all that outside of this PR if that is alright with you as this feels like a code maintainability issue that doesn't need to block this PR

**Comment by j0sh** [comment]
> Woops...yeah forgot to rebase instead of merge

It's not too late to rebase! :)

> my preference is still to do all that outside of this PR if that is alright with you.

No problem :+1:

**Review by j0sh — APPROVED** [review]

File: server/mediaserver.go
Comment (yondonfu) [reply server/mediaserver.go:229]: I was referring to the case where the `pm` package is being used outside of the Livepeer project. Though as Josh has mentioned, even if there is an session created within the `pm` package the user would still need to be aware of it somehow whether it be via a sessionID or via the returned session object itself.

**Pull request merged**
